File: mars_rover_tdd_java/src/test/java/com/example/mars_rover_tdd_java/MarsRoverTest.java
package com.example.mars_rover_tdd_java;

import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertTrue;

@SpringBootTest
class MarsRoverTest {

    @Test
    void test01_GetPosition() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        assertEquals(1, marsRover.getX());
        assertEquals(1, marsRover.getY());
        assertEquals("N", marsRover.getDirection());
    }

    @Test
    void test02_GetPosition() {
        MarsRover marsRover = new MarsRover(2,1,"N");
        assertEquals(2, marsRover.getX());
        assertEquals(1, marsRover.getY());
        assertEquals("N", marsRover.getDirection());
    }

    @Test
    void test03_IsValidPosition() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        assertTrue(marsRover.validatePosition(1, 1, "N"));
    }

    @Test
    void test04_IsNotValidPosition() {
        MarsRover marsRover = new MarsRover(2,1,"N");
        assertFalse(marsRover.validatePosition(1, 1, "N"));
    }

    @Test
    void test05_IsValidPositionWhenMoveForward() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.move("f");
        assertTrue(marsRover.validatePosition(1, 2, "N"));
    }

    @Test
    void test06_IsValidPositionWhenMoveBackward() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.move("b");
        assertTrue(marsRover.validatePosition(1, 0, "N"));
    }

    @Test
    void test07_IsValidPositionWhenTurnLeft() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.move("l");
        assertTrue(marsRover.validatePosition(1, 1, "W"));
    }

    @Test
    void test08_IsValidPositionWhenTurnRight() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.move("r");
        assertTrue(marsRover.validatePosition(1, 1, "E"));
    }

    @Test
    void test09_IsValidPositionWhenReceiveAnOrderWithCapitalLetter() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.move("F");
        assertTrue(marsRover.validatePosition(1, 2, "N"));
    }

    @Test
    void test10_IsValidPositionWhenReceiveTwoOrders() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.movementOrders("fb");
        assertTrue(marsRover.validatePosition(1, 1, "N"));
    }

    @Test
    void test11_IsValidPositionWhenReceiveThreeOrders() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.movementOrders("lfr");
        assertTrue(marsRover.validatePosition(0, 1, "N"));
    }

    @Test
    void test12_IsValidPositionWhenReceiveTenOrders() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.movementOrders("ffrfflfrff");
        assertTrue(marsRover.validatePosition(5, 4, "E"));
    }

    @Test
    void test13_IsValidPositionWhenReceiveSixOrders() {
        MarsRover marsRover = new MarsRover(1,1,"N");
        marsRover.movementOrders("fbblrr");
        assertTrue(marsRover.validatePosition(1, 0, "E"));
    }
}

File: mars_rovers_tdd_csharp/mars_rovers_tdd_csharp.test/models/MarsRoverTest.cs
using mars_rovers_tdd_csharp.models;

namespace mars_rovers_tdd_csharp.test.models;

public class MarsRoverTest
{
    [Fact]
    public void Test01_GetPosition()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        Assert.Equal(1, marsRover.GetX());
        Assert.Equal(1, marsRover.GetY());
        Assert.Equal("N", marsRover.GetDirection());
    }
    
    [Fact]
    public void Test02_GetPosition()
    {
        MarsRover marsRover = new MarsRover(2, 1, "N");
        Assert.Equal(2, marsRover.GetX());
        Assert.Equal(1, marsRover.GetY());
        Assert.Equal("N", marsRover.GetDirection());
    }
    
    [Fact]
    public void Test03_IsValidPosition()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        Assert.True(marsRover.ValidatePosition(1, 1, "N"));
    }
    
    [Fact]
    public void Test04_IsNotValidPosition()
    {
        MarsRover marsRover = new MarsRover(2, 1, "N");
        Assert.False(marsRover.ValidatePosition(1, 1, "N"));
    }
    
    [Fact]
    public void Test05_IsValidPositionWhenMoveForward()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.Move("f");
        Assert.True(marsRover.ValidatePosition(1, 2, "N"));
    }
    
    [Fact]
    public void Test06_IsValidPositionWhenMoveBackward()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.Move("b");
        Assert.True(marsRover.ValidatePosition(1, 0, "N"));
    }
    
    [Fact]
    public void Test07_IsValidPositionWhenTurnLeft()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.Move("l");
        Assert.True(marsRover.ValidatePosition(1, 1, "W"));
    }
    
    [Fact]
    public void Test08_IsValidPositionWhenTurnRight()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.Move("r");
        Assert.True(marsRover.ValidatePosition(1, 1, "E"));
    }
    
    [Fact]
    public void Test09_IsValidPositionWhenReceiveAnOrderWithCapitalLetter()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.Move("F");
        Assert.True(marsRover.ValidatePosition(1, 2, "N"));
    }
    
    [Fact]
    public void Test10_IsValidPositionWhenReceiveTwoOrders()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.MovementOrders("fb");
        Assert.True(marsRover.ValidatePosition(1, 1, "N"));
    }
    
    [Fact]
    public void Test11_IsValidPositionWhenReceiveThreeOrders()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.MovementOrders("lfr");
        Assert.True(marsRover.ValidatePosition(0, 1, "N"));
    }
    
    [Fact]
    public void Test12_IsValidPositionWhenReceiveTenOrders()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.MovementOrders("ffrfflfrff");
        Assert.True(marsRover.ValidatePosition(5, 4, "E"));
    }

    [Fact]
    public void Test13_IsValidPositionWhenReceiveSixOrders()
    {
        MarsRover marsRover = new MarsRover(1, 1, "N");
        marsRover.MovementOrders("fbblrr");
        Assert.True(marsRover.ValidatePosition(1, 0, "E"));
    }
}